lib/src/rules/invariant_booleans.dart

```diff
@@ -286,32 +161,23 @@ class _ContradictionReportRule extends LintRule {
             maturity: Maturity.stable);
 }
 
-class _Visitor extends SimpleAstVisitor {
+class _InvariantBooleansVisitor extends ConditionScopeVisitor {
   final LintRule rule;
 
-  _Visitor(this.rule);
-
-  @override
-  visitDoStatement(DoStatement node) {
-    _reportExpressionIfConstantValue(node.condition);
-  }
+  _InvariantBooleansVisitor(this.rule);
 
-  @override
-  visitForStatement(ForStatement node) {
-    _reportExpressionIfConstantValue(node.condition);
-  }
-
-  @override
-  visitIfStatement(IfStatement node) {
-    _reportExpressionIfConstantValue(node.condition);
+  TestedExpressions _findPreviousTestedExpressions(Expression node) {
+    final elements = _getRealElementsInExpression(node);
+    Iterable<Expression> conjunctions = getTrueExpressions(elements)
+        .map(_splitConjunctions)
```

Comment:
Would it be more efficient to split conjunctions when adding the conditions rather than when finding tested expressions?